Improve check for client authentication

dcooper16 opened the pull request:
As noted in #1709, some servers will only request client authentication if a specific URL is requested. This PR modifies the check for client authentication, in the case that a $URL_PATH is provided, by having testssl.sh perform a GET request on the URL provided on the command line.

I have not been able to find or set up a server that only requests client authentication if a specific URL is requested, so I have not been able to test this PR.

@lesterw1: Do you have a server that you can test this on or that you can point me to so that I can test it?

Commit:
WIP: Improve check for client authentication

As noted in #1709, some servers will only request client authentication if a specific URL is requested. This commit modifies the check for client authentication, in the case that a $URL_PATH is provided, by having testssl.sh perform a GET request on the URL provided on the command line.

drwetter commented:
Hi David,

is that still a WIP (see title?)

Thx, Dirk

dcooper16 commented:
Hi Dirk,

I haven't tried to work on this in quite a while. All that remains to be done is for it to be tested so that there is some evidence that it actually works. I haven't been able to find a server to test against or to configure Apache to only offer/require client authentication for a particular directory, and it doesn't seem that anyone else has tested it either.

drwetter commented:
yes, almost forgot that, too.

@ Anybody listening? We look for a web server which has not a general client authentication but on a specific URL only? If you have such a thing or willing to set it up, please speak up or just used @dcooper16's branch ``improve_client_auth``

drwetter commented:
Hi David,

if tested this with the following Apache config:

```
    SSLCACertificateFile /etc/ssl/client+ca.crt
    <Location /somewhere>
        SSLVerifyClient require
    </Location>
```

it returns in introducing section ``Service detected:       Couldn't determine what's running on port 443, assuming no HTTP service => skipping all HTTP checks`` if my cmd line is ``./testssl.sh https://<TAGET>somewhere``, in all other cases as usual ``HTTP``. This is independent on the openssl version (PHA support or not). It also claims (with OpenSSL 1.1.1x in the server defaults section) `` Client Authentication        none``

If I remove the location statements 

```
    SSLCACertificateFile /etc/ssl/client+ca.crt
        SSLVerifyClient require
```

both work as expected and as before.

drwetter commented:
thanks for your PR btw, and sorry for the wait

Commit:
Improve check for client authentication

Based on initial testing, this commit improves the check for client authentication in the case that the server only requests client authentication for specific URLs. However, it does not work correctly if the server supports TLS 1.3 and $OPENSSL is a version of LibreSSL that supports TLS 1.3 in s_client. The problem is that LibreSSL does not support post-handshake authentication with TLS 1.3

dcooper16 commented:
Hi Dirk,

Thanks for the information. When I tried configuring Apache I was using `<Directory /somewhere>` rather than `<Location /somewhere>`. After changing "Directory" to "Location" I was able to do some testing and after trying a few different things found something that seems to work, although it needs to be tested some more.

There is definitely one problem with this PR. If `$OPENSSL` is a version of LibreSSL for which s_client supports TLS 1.3 and the server supports TLS 1.3, then the results will be incorrect. The problem is that LibreSSL's s_client does not seem to support post-handshake authentication, so there is no way to get the server to send a `CertificateRequest` message.

drwetter commented:
Hi David,

thanks for the feedback. Let me know if I can help you, like e.g. setting up a machine out there (=aws) for you. I did my tests on an in internal machine.

Thx, Dirk

Commit:
Improve check for client authentication with LibreSSL

Checking for client authentication with TLS 1.3 requires post-handshake authentication, which does not appear to be supported by LibreSSL. This commit improves the check for client authentication when testing a TLS 1.3 server using LibreSSL by having determine_optimal_proto() first test for connectivity with TLS 1.3 without checking for client authentication and then performing a separate check for client authentication using a non-TLS 1.3 protocol.

This commit only affects the flow of the program if a $URL_PATH is specified, the server supports TLS 1.3, and $OPENSSL supports TLS 1.3 but not -enable_pha.

testss.sh may still provide incorrect information about client authentication if a $URL_PATH is provided, the server is TLS 1.3-only, and LibreSSL is used.

Commit:
Report if couldn't test for client authentication

This commit fixes determine_optimal_proto() and run_server_defaults() so that a "Local problem" is reported if a $URL_PATH is specified, the server is TLS 1.3-only, and $OPENSSL does not support -enable_pha (and the server does not offer client authentication as part of the initial TLS handshake).

Commit:
Fix potential stallling in HTTP query

In run_http_header() the GET command is first sent over TLS using a background process, and then if that does not hang, it is sent again in the foreground. Similarly, service_detection() runs the command in the background.

This commit changes determine_optimal_proto() to follow the example of run_http_header() as protection against the possibility of the HTTP query stalling.

dcooper16 commented:
I've made some changes to this PR so that it works better when using testssl.sh with LibreSSL against a server that supports TLS 1.3. I also modified the code to try to protect against `$OPENSSL` hanging when sending a GET request through it (following the example from `run_http_header()`), just to be safe.

So, at this point I think this PR is ready to be merged,

drwetter commented:
Works for me now and looks great.

Thanks a lot, David!
